modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheImpl.java

```diff
@@ -289,7 +259,14 @@ public BucketList getBucketList(Experiment.ID expId) {
             bucketList = (BucketList) val.getObjectValue();
         }
 
-        return isNull(bucketList)?new BucketList():bucketList;
+        return isNull(bucketList)?new BucketList():makeCopy(bucketList);
+    }
+
+    private BucketList makeCopy(BucketList bucketList) {
+        List<Bucket> iBucketList = new ArrayList<>(bucketList.getBuckets());
```

Comment:
Making a copy of BucketList to avoid ConcurrentModificationException... At some point during the assignment logic, the List<Bucket> object inside this BucketList is getting sorted, I believe that's the reason of ConcurrentModificationException. 

AssignmentImpl.selectBucket(List<Bucket> buckets) Line#1189.

Provided comments to this method...